### PR TITLE
Libcloud 868 pdns 4

### PR DESCRIPTION
**Pull request by biggosh**
### Description

Changed create_record, update_record, to_records, _to_record to reflect changes in REST API specifications
### Status
- done, ready for review
### Checklist

Tested vs latest PDNS version 4.0.1 on a ubuntu server 16.10. It was installed from ubuntu repository

### CODE REVIEW

File: libcloud/dns/drivers/powerdns.py [hunk 1]
```diff
@@ -248,18 +258,18 @@ def create_zone(self, domain, type=None, ttl=None, extra={}):
...
-                                             value=e.message)
+            if e.code == httplib.UNPROCESSABLE_ENTITY and\
+                e.message.startswith("Domain '%s' already exists" % domain):
+                raise ZoneAlreadyExistsError(zone_id = zone_id, driver = self,
+                                             value = e.message)
             raise e
-        return Zone(id=zone_id, domain=domain, type=None, ttl=None,
-                    driver=self, extra=extra)
-
+        return Zone(id = zone_id, domain = domain, type = None, ttl = None,
```
Comment (Kami) [inline libcloud/dns/drivers/powerdns.py]: Can you please revert all the style changes you have done? They don't comply with pep8 we follow and also make the pull request hard to review (hard to differentiate between actual code and style changes).

Thanks!

**Comment by biggosh** [comment]
I hope I removed all the formatting.
May you check? 

thanks

**Comment by tonybaloney** [comment]
hey @biggosh thanks for the contribution, you'll see there's still a red X next to the commit, which means the build failed https://travis-ci.org/apache/libcloud/builds/171641860

Check out the instructions on testing - http://libcloud.readthedocs.io/en/latest/testing.html

I normally at least run `tox -e py2.7` and `tox -e py3.5`

**Comment by biggosh** [comment]
I've modified expected response from powerdns in accordance to API REST 4.0 specs.
Now functional tests pass, I've noticed one test fails on coding style.

**Comment by biggosh** [comment]
Are there any other changes needed? I'm using the code changed on one my project and it works; so also real every day use confirms tests result.

thanks

**Comment by biggosh** [comment]
Any news about this pull request?

**Comment by pquentin** [comment]
Sorry, we don't have enough reviewers and it has been forgotten. Do you think you could rebase this on top of trunk? Sorry about this!

**Comment by biggosh** [comment]
I've rebased. I'm not updating the code since 2016 when I used it for one of my projects.
Let me know if some more work is needed

**Comment by pquentin** [comment]
Thanks! Appreciated.

Can you please fix the lint? See https://travis-ci.org/apache/libcloud/jobs/337021710 and `tox -e lint`.

**Comment by biggosh** [comment]
Done.

File: libcloud/dns/drivers/powerdns.py [hunk 2]
```diff
@@ -184,17 +190,22 @@ def create_record(self, name, zone, type, data, extra=None):
...
+            'rrsets': [{
+                'name': name,
+                'type': type,
+                'changetype': 'REPLACE',
+                'ttl': extra['ttl'],
+                'records': [record],
+            }]
+        }
+
+        if (comment):
```
Comment (pquentin) [inline libcloud/dns/drivers/powerdns.py:206]: `if comment:` is enough

File: libcloud/dns/drivers/powerdns.py [hunk 3]
```diff
@@ -403,21 +424,22 @@ def update_record(self, record, name, type, data, extra=None):
...
+            'rrsets': [{
+                'name': name,
+                'type': type,
+                'changetype': 'REPLACE',
+                'ttl': extra['ttl'],
+                'records': [updated_record],
+            }]
+        }
+
+        if (comment):
```
Comment (pquentin) [inline libcloud/dns/drivers/powerdns.py:440]: same as above

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks! We're nearly there. Can you please use the `comment` parameter in tests so that those lines are covered? We don't accept pull requests that make coverage go down.

Sorry for my brevity in the comments below.

**Comment by biggosh** [comment]
may you explain better? where and what I've to add.

**Comment by pquentin** [comment]
It would be in libcloud/test/dns/test_powerdns.py in the `test_create_record` function, and it's possible that fixtures in libcloud/test/dns/fixtures/powerdns/ would have to change.

However I'm worried here that we're changing the behavior for all PowerDNS versions (3 and 4) while we would like to only change the behavior for PowerDNS 4.

@ktdreyer can you please help to review this pull request? I'm trying to help but never used the DNS parts of libcloud. Thanks!

**Comment by pquentin** [comment]
@biggosh Please make sure tests pass before submitting new commits: see http://libcloud.readthedocs.io/en/latest/testing.html. Lint should pass, you can test this with `tox -e lint`.

Do you have any thoughts about breaking PowerDNS 3.0 ? We should either deprecate the 3.0 API (but it's currently the default version), or only introduce this change for PowerDNS 4.0.

**Comment by biggosh** [comment]
Maybe I found a partial solution not to brake 3.0 compatibility. I don't
have a test environment with it, It will take time to me to prepare one.

In next days I'll solve issues I introduced yesterday night.

**Comment by pquentin** [comment]
@biggosh Anything I can help with?

**Comment by biggosh** [comment]
I'll restart working on it next week.

**Comment by Kami** [comment]
First thank you for your contribution.

This PR is very old and out of date now so I will go ahead and close it.

If it's still relevant, please sync your changes with the latest master / trunk and open a new PR. Thanks!

**Pull request closed without merging**